### ISSUE TITLE
[FEATURE] Improving agents by robust prompt compilation and optimization

### ISSUE THREAD

**Issue by supreme-gg-gg** [kagent-dev/kagent#981]
### 📋 Prerequisites

- [x] I have searched the [existing issues](./issues) to avoid creating a duplicate
- [x] By submitting this issue, you agree to follow our [Code of Conduct](https://github.com/kagent-dev/kagent/blob/main/CODE_OF_CONDUCT.md)

### 📝 Feature Summary

Improve agents by integrating the DSPy framework as a native prompt compilation and optimization service. This will allow users to declaratively create prompts and automatically compile them into robust, efficient prompts, solving the challenge of manual prompt engineering for production systems, while allowing for them to be optimised with DSPy's algorithms.

### ❓ Problem Statement / Motivation

Hand-crafted prompts are a major bottleneck for deploying reliable agents in production. Yet, most agents on Kagent (declarative) are created with this in design (the user types in a big string into the input box). The problems with writing (and hand-optimising) prompts like this are:

- Poor Generalization: Overfit to development examples, unpredictable for edge cases, fails in production
- Manual & Inefficient: Prompt engineering is analogous to manual, low-throughput, non-systematic hyperparamter tuning in ML
- Model-Specific Coupling: Prompts are tightly coupled to the specific LLM they were designed for

This feature solves the problem of prompt fragility and maintenance by replacing manual tuning with a systematic, automated optimization process. It would benefit any developer on our platform aiming to build agents that are reliable, maintainable, and adaptable to the evolving LLM landscape.

### 💡 Proposed Solution

I propose integrating [DSPy](https://github.com/stanfordnlp/dspy), a research-driven, production-tested toolkit for programming—not just prompting—LLMs.

By adding DSPy as a native prompt optimization service in Kagent, we empower users to automatically transform high-level program definitions into robust, optimized prompts. This decouples the difficult task of prompt engineering from the agent's core orchestration logic.

The proposed user workflow would be:

1. Program the LLM: Users define their task declaratively by specifying a signature (inputs/outputs) and a module (e.g., ChainOfThought, ReAct), which can be exposed through a simple UI.

2. Deploy as a Reusable Prompt: The compiled prompt artifact is saved and versioned, integrating seamlessly with our existing prompt management infrastructure for ConfigMaps and Secrets (as introduced in #894).

3. One-Click Optimization: A user triggers an optimization job if needed. A DSPy optimizer (e.g., GEPA, MIPRO) runs in the background, using an evaluation set / user-defined metrics / teacher model to find the most effective prompt for the target LLM. **Having agent evaluation implemented would help with this step, but it is not strictly necessary**

**Note that this feature request is completely independent from the DSPy framework itself. I am happy to integrate any other framework that supports prompt optimization and compilation.**

### 🔄 Alternatives Considered

- Status Quo (Manual Prompt Engineering): Continue requiring users to write prompts manually within our declarative agent framework. This is the direct cause of the fragility and high maintenance costs we are trying to solve.

- User-Managed Integration: Force users to integrate libraries like DSPy themselves in "Bring Your Own" (BYO) agent containers. This creates a high barrier to entry and fails to provide any benefit to users of our declarative agent system.

- External Management: Allow users to paste prompts from an external source. This is unmanageable at scale, as it lacks versioning, reusability, and a systematic workflow.

- Fine-tuning: The other popular method to make agents perform better is through reinforcement learning, using frameworks such as Verl or (ART)[https://github.com/openpipe/art] built on top of GRPO-like algorithms. However, prompt optimization is much more beginner friendly and requires little effort, time, and compute in comparison, while being shown to outperform RL in certain cases (see [GEPA paper](https://arxiv.org/pdf/2507.19457))

### 🎯 Affected Service(s)

None

### 📚 Additional Context

Google Vertex AI already have a similar feature: https://cloud.google.com/vertex-ai/generative-ai/docs/learn/prompts/prompt-optimizer

This feature does not seem available in most agent frameworks like ADK, LangGraph, CrewAI. I am not sure if we should wait until the framework releases these integration. However, my argument here would be that these frameworks excel at orchestrating workflows, but micro-optimising the prompts at each step could be "refactored" out into an independent service.

Another point to note is that this is a significant performance booster for those using smaller language models (e.g. Llama, Gemma, Phi, Qwen) for their agents.

If there is interest in this feature, I will go ahead and create a more comprehensive design document.

### 🙋 Are you willing to contribute?

- [x] I am willing to submit a PR for this feature

**Comment by peterj** [comment]
This would definitely be an interesting improvement -- I am not familiar with DSPy, but would this be something that's part of the UI where instead of users entering the prompts directly, they would launch into the new UI that takes the inputs/output and models (as you mentioned) and the output of that would be an optimized prompt? Or is this a longer process that takes more time (involves evaluation sets, user metrics, etc.)?

The reusable prompts are interesting as well, but could also be implemented separately (regardless if we use this optimization or not).

**Comment by supreme-gg-gg** [comment]
The prompt compilation will be pretty much instant, the optimization can either be sync or async depending on how long it takes. 

The UI can allow the user to define these modular, high-level components such as input / output (question -> answer), modules (ReAct for example), tools, etc. and the framework like DSPy will compile these into the prompt. There can be a toggle on the UI if the user wants to use this interface or the traditional string input box.

Regarding optimization, the following diagram borrowed from langgraph / langsmith might help explain (the optimizer algorithm itself is customisable, examples can also be synthetic or user provided)

<img width="683" height="341" alt="Image" src="https://github.com/user-attachments/assets/064487bd-c1c6-46eb-99e9-153e138df3a4" />

Source: https://langchain-ai.github.io/langmem/reference/prompt_optimization/

This involves several LLM calls to large models, so making it async might be better. I don't think ADK has any prompt optimization tools, I will probably experiment with DSPy + ADK first and update on the issue if there's anything fruitful.

**Comment by rishabh998186** [comment]
I am willing to submit a PR for this feature

**Comment by supreme-gg-gg** [comment]
> I am willing to submit a PR for this feature

After investigating a bit more and chatting with several people from the DSPy community, I do not think this feature is as straightforward as I thought. One of the discussion thread is here: https://github.com/stanfordnlp/dspy/issues/8883. It might still be possible to integrate DSPy the way I proposed, but it would not be as optimised and simple.

I think more thought needs to go into this issue before any implementation is made. However, we don't necessarily need to stick with DSPy, perhaps having just a simpler prompt optimizer would be great. The langgraph one linked above is a great reference (with this blog as elaboration as well: https://blog.langchain.com/exploring-prompt-optimization/). We can consider something similar.

**Comment by rishabh998186** [comment]
Hey @supreme-gg-gg!

You're absolutely right about this being way too complex. After reading your comment and that LangChain post, I realized I might have gone overboard with the whole DSPy thing.

So here's the deal - I did implement the full DSPy microservice with Python server and all that jazz (But honestly, this is quite a large codebase). Then I read your feedback and thought "man, this is exactly what he was warning about".


After seeing what you suggested about simpler prompt optimization, I also built a lightweight version that just does the core stuff:

Takes a prompt

Generates some variations

Tests them with A/B approach

Picks the best one

Done in like ~200 lines

The simple version basically does what users actually want - "make my prompts better" - without all the overhead of running separate services and dealing with DSPy's complexity.

What I'm thinking is:

Scrap the whole DSPy microservice thing

Keep just the simple optimizer that runs in-process

Focus on that core user value you mentioned

The simple approach would literally cut this down from 2000+ lines to maybe 150-200 lines. Same result for users, way less headache for maintenance.

Should I just go with the simple approach and ditch all the DSPy stuff? I feel like that's what would actually be useful without being a pain to deploy and maintain.

Let me know what you think!             

<img width="1920" height="1080" alt="Image" src="https://github.com/user-attachments/assets/48b32e13-b609-4297-a549-52919265c53b" />
<img width="1920" height="1080" alt="Image" src="https://github.com/user-attachments/assets/601cd400-0a82-49ee-afc9-695f432d56f2" />
<img width="1920" height="1080" alt="Image" src="https://github.com/user-attachments/assets/dae4bf53-2530-442e-a755-9978be95bf1f" />
<img width="1920" height="1080" alt="Image" src="https://github.com/user-attachments/assets/65b9adb2-a192-4e47-af58-5bcff1b7d6fb" />

**Comment by supreme-gg-gg** [comment]
Impressive effort that you got DSPy integration working 👀! How does the current approach work, does it process the prompt with DSPy and create some reusable resource for the optimised prompt? How is it gathering trajectories and feedback right now? I personally think simpler approach might just suffice, and A/B testing could be a good place to start as you suggested. We can evolve this after the agent evaluation system is added perhaps (because most gradient and meta-prompting based techniques require evaluation to be setup)

**Comment by rishabh998186** [comment]
when a user toggles on "DSPy Mode" in the UI, their prompt gets processed through various DSPy reasoning modules like ChainOfThought and Predict, then it gets stored in the database as an optimized prompt resource that the agent can use going forward. The trajectory and feedback collection part is where things get tricky though - right now I've got the database schema set up for storing optimization jobs and some basic telemetry stubs, but the actual feedback loop isn't really implemented yet because you're totally right, it would need a proper evaluation system to work effectively.

**Comment by rishabh998186** [comment]
lets keep it simple for now.You are right that the fancy optimization stuff needs proper evaluation anyway, so no point building that yet  so I focus on the A/B testing approach for now also  My plan is to build out the simple genetic optimizer that generates prompt variations, tests them with sample queries, and picks the best one - all happening in-process without any external dependencies. Once the evaluation system is in place, we can definitely evolve this into the more sophisticated techniques since the interface will already be there.

**Comment by supreme-gg-gg** [comment]
> lets keep it simple for now.You are right that the fancy optimization stuff needs proper evaluation anyway, so no point building that yet so I focus on the A/B testing approach for now also My plan is to build out the simple genetic optimizer that generates prompt variations, tests them with sample queries, and picks the best one - all happening in-process without any external dependencies. Once the evaluation system is in place, we can definitely evolve this into the more sophisticated techniques since the interface will already be there.

Agreed. Evaluation should come soon (I'm working on it 😄) I'll keep this in mind when drafting the design
